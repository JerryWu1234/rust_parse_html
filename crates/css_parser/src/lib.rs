#[derive(Debug)]
pub struct Stylesheet {
  pub rules: Vec<Rule>,
}

#[derive(Debug)]
pub struct Rule {
  pub selectors: Vec<Selector>,
  pub declarations: Vec<Declaration>,
}

#[derive(Debug)]
pub enum Selector {
  Simple(SimpleSelector),
}

#[derive(Debug)]
pub struct SimpleSelector {
  pub tag_name: Option<String>,
  pub id: Option<String>,
  pub class: Vec<String>,
}

#[derive(Debug)]
pub struct Declaration {
  pub name: String,
  pub value: Value,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Value {
  Keyword(String),
  Length(f32, Unit),
  ColorValue(Color),
}

#[derive(Debug, Clone, PartialEq)]
pub enum Unit {
  Px,
}

#[derive(Debug, Clone, PartialEq, Default)]
pub struct Color {
  pub r: u8,
  pub g: u8,
  pub b: u8,
  pub a: u8,
}
pub fn parse(source: String) -> Stylesheet {
  let mut parser = Parser {
    pos: 0,
    input: source,
  };
  Stylesheet {
    rules: parser.parse_rules(),
  }
}
pub struct Parser {
  pub pos: usize,
  pub input: String,
}

impl Selector {
  pub fn speificity(&self) -> (usize, usize, usize) {
    let Selector::Simple(valeu) = self;
    let a = valeu.id.iter().count();
    let v = valeu.class.len();
    let c = valeu.tag_name.iter().count();
    (a, v, c)
  }
}

impl Parser {
  pub fn parse_rules(&mut self) -> Vec<Rule> {
    let mut sheet: Vec<Rule> = vec![];
    loop {
      self.consume_whitespace();
      if self.eof() {
        break;
      }
      sheet.push(self.parse_rule())
    }
    sheet
  }
  pub fn parse_rule(&mut self) -> Rule {
    Rule {
      selectors: self.parse_selectors(),
      declarations: self.parse_declarations(),
    }
  }

  pub fn parse_declarations(&mut self) -> Vec<Declaration> {
    assert_eq!(self.consume_char(), '{');
    let mut declarations: Vec<Declaration> = vec![];
    loop {
      self.consume_whitespace();
      if self.next_char() == '}' {
        self.consume_char();
        break;
      }
      declarations.push(self.parse_declaration());
    }
    declarations
  }

  pub fn parse_declaration(&mut self) -> Declaration {
    let name = self.iteration(getValue);
    self.consume_whitespace();
    assert_eq!(self.consume_char(), ':');
    self.consume_whitespace();
    let value = self.parse_value();
    self.consume_whitespace();
    assert_eq!(self.consume_char(), ';');
    Declaration {
      name: name,
      value: value,
    }
  }

  pub fn parse_value(&mut self) -> Value {
    match self.next_char() {
      '0'..='9' => self.parse_length(),
      '#' => self.parse_color(),
      _ => Value::Keyword(self.iteration(getValue)),
    }
  }

  pub fn parse_color(&mut self) -> Value {
    assert_eq!(self.consume_char(), '#');
    Value::ColorValue(Color {
      r: self.parse_hex_pair(),
      g: self.parse_hex_pair(),
      b: self.parse_hex_pair(),
      a: 255,
    })
  }

  pub fn parse_hex_pair(&mut self) -> u8 {
    let s = &self.input[self.pos..self.pos + 2];
    self.pos += 2;
    u8::from_str_radix(s, 16).unwrap()
  }

  pub fn parse_length(&mut self) -> Value {
    let value = self.iteration(|c| match c {
      '0'..='9' => true,
      _ => false,
    });

    let unit = self.iteration(getValue).to_ascii_lowercase();

    Value::Length(
      value.parse().unwrap(),
      match &*unit {
        "px" => Unit::Px,
        _ => panic!("unrecognized unit"),
      },
    )
  }
  pub fn parse_selectors(&mut self) -> Vec<Selector> {
    let mut selectors = vec![];
    loop {
      selectors.push(Selector::Simple(self.parse_simple_selector()));
      self.consume_whitespace();
      match self.next_char() {
        ',' => {
          self.consume_char();
          self.consume_whitespace();
        }
        '{' => break,
        c => panic!("Unexpected character {} in selector list", c),
      }
    }
    selectors.sort_by(|a, b| b.speificity().cmp(&a.speificity()));
    selectors
  }

  pub fn parse_simple_selector(&mut self) -> SimpleSelector {
    let mut selector = SimpleSelector {
      tag_name: None,
      id: None,
      class: vec![],
    };
    while !self.eof() {
      match self.next_char() {
        '#' => {
          self.consume_char();
          selector.id = Some(self.iteration(getValue))
        }
        '.' => {
          self.consume_char();
          selector.class.push(self.iteration(getValue))
        }
        '*' => {
          self.consume_char();
        }
        c if getValue(c) => selector.tag_name = Some(self.iteration(getValue)),
        _ => break,
      }
    }
    selector
  }

  pub fn next_char(&mut self) -> char {
    self.input[self.pos..].chars().next().unwrap()
  }

  pub fn consume_whitespace(&mut self) {
    self.iteration(char::is_whitespace);
  }

  pub fn iteration<F>(&mut self, callback: F) -> String
  where
    F: Fn(char) -> bool,
  {
    let mut value = String::from("");
    while !self.eof() && callback(self.next_char()) {
      value.push(self.consume_char())
    }
    value
  }

  pub fn consume_char(&mut self) -> char {
    let (index, char) = self.input[self.pos..].char_indices().next().unwrap();
    self.pos += index + 1;
    println!("current_char:{}", char);
    char
  }
  pub fn eof(&mut self) -> bool {
    self.pos >= self.input.len()
  }
}

fn getValue(c: char) -> bool {
  match c {
    'a'..='z' | 'A'..='Z' | '0'..='9' | '-' | '_' => true,
    _ => false,
  }
}
