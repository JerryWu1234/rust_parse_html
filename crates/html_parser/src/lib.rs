pub mod dom;
// use std::{collections::HashMap, hash::Hash};

use std::collections::HashMap;

pub use dom::*;

#[derive(Debug)]
pub struct Parser {
    pub pos: usize,
    pub input: String,
}

pub fn parse(source: String) -> dom::Node {
    let mut nodes = Parser {
        pos: 0,
        input: source,
    }
    .parse_nodes();

    // If the document contains a root element, just return it. Otherwise, create one.
    if nodes.len() == 1 {
        nodes.swap_remove(0)
    } else {
        dom::create_element("html".to_string(), HashMap::new(), nodes)
    }
}

impl Parser {
    pub fn parse_nodes(&mut self) -> Vec<Node> {
        let mut node: Vec<Node> = vec![];
        loop {
            self.consume_whitespace();
            // println!(">>{:?}", self.input[self.pos..].to_string());
            println!("{:#?}", self.input[self.pos..].to_string());
            if self.eof() || self.input[self.pos..].starts_with("</") {
                break;
            }
            node.push(self.parse_node())
        }
        node
    }

    pub fn parse_node(&mut self) -> Node {
        match self.next_char() {
            '<' => self.cusume_element(),
            _ => self.cusume_text(),
        }
    }

    pub fn cusume_text(&mut self) -> Node {
        dom::create_text(self.iteration(|c| c != '<'))
    }

    pub fn cusume_element(&mut self) -> Node {
        assert_eq!('<', self.consume_char());
        let tag_name = self.tag_name();
        self.consume_whitespace();
        let attributes = self.consume_attributes();
        assert_eq!('>', self.consume_char());
        let children = self.parse_nodes();
        assert_eq!(self.consume_char(), '<');
        assert_eq!(self.consume_char(), '/');
        assert_eq!(self.tag_name(), tag_name);
        assert_eq!(self.consume_char(), '>');
        dom::create_element(tag_name, attributes, children)
    }

    pub fn consume_attributes(&mut self) -> HashMap<String, String> {
        let mut attrs = HashMap::new();
        loop {
            self.consume_whitespace();
            if self.next_char() == '>' {
                break;
            }
            let (key, value) = self.get_attributes();
            attrs.insert(key, value);
        }
        attrs
    }

    pub fn get_attributes(&mut self) -> (String, String) {
        let key = self.iteration(getStrings);
        assert_eq!('=', self.consume_char());
        let open_quote = self.consume_char();
        assert!(open_quote == '"' || open_quote == '\'');
        let value = self.iteration(|c| c != open_quote);
        assert_eq!('"', self.consume_char());

        (key, value)
    }

    pub fn tag_name(&mut self) -> String {
        self.iteration(getStrings)
    }

    pub fn consume_whitespace(&mut self) {
        self.iteration(char::is_whitespace);
    }

    pub fn next_char(&mut self) -> char {
        self.input[self.pos..].chars().next().unwrap()
    }

    pub fn iteration<F>(&mut self, callback: F) -> String
    where
        F: Fn(char) -> bool,
    {
        let mut value = String::from("");
        while !self.eof() && callback(self.next_char()) {
            value.push(self.consume_char())
        }
        value
    }

    pub fn consume_char(&mut self) -> char {
        let (index, char) = self.input[self.pos..].char_indices().next().unwrap();
        self.pos += 1;
        println!("{},>>>打印当前循环的char{}", char, index);
        char
    }

    pub fn eof(&self) -> bool {
        self.pos >= self.input.len()
    }
}

fn getStrings(v: char) -> bool {
    match v {
        'a'..='z' | 'A'..='Z' | '0'..='9' => true,
        _ => false,
    }
}
