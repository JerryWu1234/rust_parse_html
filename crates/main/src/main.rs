use std::{
  fs::File,
  io::{BufWriter, Read},
};
mod paint;
extern crate image;

use css_parser;
use html_parser;

fn main() {
  // Parse command-line options:
  // let mut opts = getopts::Options::new();
  // opts.optopt("h", "html", "HTML document", "FILENAME");
  // opts.optopt("c", "css", "CSS stylesheet", "FILENAME");
  // opts.optopt("o", "output", "Output file", "FILENAME");
  // opts.optopt("f", "format", "Output file format", "png | pdf");

  // let matches = opts.parse(std::env::args().skip(1)).unwrap();
  // let str_arg = |flag: &str, default: &str| -> String {
  //     matches.opt_str(flag).unwrap_or(default.to_string())
  // };

  // // Choose a format:
  // let png = match &str_arg("f", "png")[..] {
  //     "png" => true,
  //     "pdf" => false,
  //     x => panic!("Unknown output format: {}", x),
  // };

  // Read input files:
  let html = read_source("example/test.html".to_string());
  let css = read_source("example/test.css".to_string());

  // Since we don't have an actual window, hard-code the "viewport" size.
  let mut viewport: layout::Dimensions = Default::default();
  viewport.content.width = 800.0;
  viewport.content.height = 600.0;

  // Parsing and rendering:
  let root_node = html_parser::parse(html);
  let stylesheet = css_parser::parse(css);
  let style_root = style::style_tree(&root_node, &stylesheet);
  let layout_root = layout::layout_tree(&style_root, viewport);
  // println!(">>>>>>>>{:#?}", layout_root);
  // Create the output file:
  // let filename = str_arg("o", if png { "output.png" } else { "output.pdf" });
  let mut file = BufWriter::new(File::create("output.png").unwrap());

  // Write to the file:
  let ok = {
    let canvas = paint::paint(&layout_root, viewport.content);
    let (w, h) = (canvas.width as u32, canvas.height as u32);
    let img = image::ImageBuffer::from_fn(w, h, move |x, y| {
      let color = canvas.pixels[(y * w + x) as usize];
      image::Pixel::from_channels(color.r, color.g, color.b, color.a)
    });
    image::ImageRgba8(img).save(&mut file, image::PNG).is_ok()
  };
  if ok {
    println!("Saved output as {}", "output")
  } else {
    println!("Error saving output as {}", "output")
  }
}

fn read_source(filename: String) -> String {
  let mut str = String::new();
  File::open(filename)
    .unwrap()
    .read_to_string(&mut str)
    .unwrap();
  str
}

#[test]
fn test2() {
  use insta::assert_debug_snapshot;
  let html = read_source("example/test.html".to_string());
  let css = read_source("example/test.css".to_string());

  // Since we don't have an actual window, hard-code the "viewport" size.
  let mut viewport: layout::Dimensions = Default::default();
  viewport.content.width = 800.0;
  viewport.content.height = 600.0;

  // Parsing and rendering:
  let root_node = html_parser::parse(html);
  let stylesheet = css_parser::parse(css);
  let style_root = style::style_tree(&root_node, &stylesheet);
  assert_debug_snapshot!(style_root);
  // let layout_root = layout::layout_tree(&style_root, viewport);
  // assert_debug_snapshot!(layout_root);
}


#[test]
fn test3() {
  use insta::assert_debug_snapshot;
  let html = read_source("example/test.html".to_string());
  let css = read_source("example/test.css".to_string());

  // Since we don't have an actual window, hard-code the "viewport" size.
  let mut viewport: layout::Dimensions = Default::default();
  viewport.content.width = 800.0;
  viewport.content.height = 600.0;

  // Parsing and rendering:
  let root_node = html_parser::parse(html);
  let stylesheet = css_parser::parse(css);
  let style_root = style::style_tree(&root_node, &stylesheet);
  // assert_debug_snapshot!(style_root);
  let layout_root = layout::layout_tree(&style_root, viewport);
  assert_debug_snapshot!(layout_root);
}