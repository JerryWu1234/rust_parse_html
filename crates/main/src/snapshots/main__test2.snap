---
source: crates/main/src/main.rs
assertion_line: 91
expression: style_root

---
StyledNode {
    node: Node {
        children: [
            Node {
                children: [
                    Node {
                        children: [
                            Node {
                                children: [],
                                node_type: Text(
                                    "Test",
                                ),
                            },
                        ],
                        node_type: Element(
                            ElementPreproty {
                                tag_name: "title",
                                attributes: {},
                            },
                        ),
                    },
                ],
                node_type: Element(
                    ElementPreproty {
                        tag_name: "head",
                        attributes: {},
                    },
                ),
            },
            Node {
                children: [
                    Node {
                        children: [
                            Node {
                                children: [],
                                node_type: Text(
                                    "Hello, ",
                                ),
                            },
                            Node {
                                children: [
                                    Node {
                                        children: [],
                                        node_type: Text(
                                            "world!",
                                        ),
                                    },
                                ],
                                node_type: Element(
                                    ElementPreproty {
                                        tag_name: "span",
                                        attributes: {
                                            "id": "name",
                                        },
                                    },
                                ),
                            },
                        ],
                        node_type: Element(
                            ElementPreproty {
                                tag_name: "p",
                                attributes: {
                                    "class": "inner",
                                },
                            },
                        ),
                    },
                    Node {
                        children: [
                            Node {
                                children: [],
                                node_type: Text(
                                    "Goodbye!\n    ",
                                ),
                            },
                        ],
                        node_type: Element(
                            ElementPreproty {
                                tag_name: "p",
                                attributes: {
                                    "class": "inner",
                                    "id": "bye",
                                },
                            },
                        ),
                    },
                ],
                node_type: Element(
                    ElementPreproty {
                        tag_name: "div",
                        attributes: {
                            "class": "outer",
                        },
                    },
                ),
            },
        ],
        node_type: Element(
            ElementPreproty {
                tag_name: "html",
                attributes: {},
            },
        ),
    },
    specified_values: {
        "border-width": Length(
            1.0,
            Px,
        ),
        "margin": Keyword(
            "auto",
        ),
        "display": Keyword(
            "block",
        ),
        "padding": Length(
            10.0,
            Px,
        ),
        "background": ColorValue(
            Color {
                r: 255,
                g: 255,
                b: 255,
                a: 255,
            },
        ),
        "width": Length(
            600.0,
            Px,
        ),
    },
    children: [
        StyledNode {
            node: Node {
                children: [
                    Node {
                        children: [
                            Node {
                                children: [],
                                node_type: Text(
                                    "Test",
                                ),
                            },
                        ],
                        node_type: Element(
                            ElementPreproty {
                                tag_name: "title",
                                attributes: {},
                            },
                        ),
                    },
                ],
                node_type: Element(
                    ElementPreproty {
                        tag_name: "head",
                        attributes: {},
                    },
                ),
            },
            specified_values: {
                "display": Keyword(
                    "none",
                ),
            },
            children: [
                StyledNode {
                    node: Node {
                        children: [
                            Node {
                                children: [],
                                node_type: Text(
                                    "Test",
                                ),
                            },
                        ],
                        node_type: Element(
                            ElementPreproty {
                                tag_name: "title",
                                attributes: {},
                            },
                        ),
                    },
                    specified_values: {
                        "display": Keyword(
                            "block",
                        ),
                    },
                    children: [
                        StyledNode {
                            node: Node {
                                children: [],
                                node_type: Text(
                                    "Test",
                                ),
                            },
                            specified_values: {},
                            children: [],
                        },
                    ],
                },
            ],
        },
        StyledNode {
            node: Node {
                children: [
                    Node {
                        children: [
                            Node {
                                children: [],
                                node_type: Text(
                                    "Hello, ",
                                ),
                            },
                            Node {
                                children: [
                                    Node {
                                        children: [],
                                        node_type: Text(
                                            "world!",
                                        ),
                                    },
                                ],
                                node_type: Element(
                                    ElementPreproty {
                                        tag_name: "span",
                                        attributes: {
                                            "id": "name",
                                        },
                                    },
                                ),
                            },
                        ],
                        node_type: Element(
                            ElementPreproty {
                                tag_name: "p",
                                attributes: {
                                    "class": "inner",
                                },
                            },
                        ),
                    },
                    Node {
                        children: [
                            Node {
                                children: [],
                                node_type: Text(
                                    "Goodbye!\n    ",
                                ),
                            },
                        ],
                        node_type: Element(
                            ElementPreproty {
                                tag_name: "p",
                                attributes: {
                                    "class": "inner",
                                    "id": "bye",
                                },
                            },
                        ),
                    },
                ],
                node_type: Element(
                    ElementPreproty {
                        tag_name: "div",
                        attributes: {
                            "class": "outer",
                        },
                    },
                ),
            },
            specified_values: {
                "border-width": Length(
                    2.0,
                    Px,
                ),
                "border-color": ColorValue(
                    Color {
                        r: 102,
                        g: 102,
                        b: 102,
                        a: 255,
                    },
                ),
                "background": ColorValue(
                    Color {
                        r: 0,
                        g: 204,
                        b: 255,
                        a: 255,
                    },
                ),
                "padding": Length(
                    50.0,
                    Px,
                ),
                "display": Keyword(
                    "block",
                ),
                "margin": Length(
                    50.0,
                    Px,
                ),
            },
            children: [
                StyledNode {
                    node: Node {
                        children: [
                            Node {
                                children: [],
                                node_type: Text(
                                    "Hello, ",
                                ),
                            },
                            Node {
                                children: [
                                    Node {
                                        children: [],
                                        node_type: Text(
                                            "world!",
                                        ),
                                    },
                                ],
                                node_type: Element(
                                    ElementPreproty {
                                        tag_name: "span",
                                        attributes: {
                                            "id": "name",
                                        },
                                    },
                                ),
                            },
                        ],
                        node_type: Element(
                            ElementPreproty {
                                tag_name: "p",
                                attributes: {
                                    "class": "inner",
                                },
                            },
                        ),
                    },
                    specified_values: {
                        "width": Length(
                            500.0,
                            Px,
                        ),
                        "margin-bottom": Length(
                            20.0,
                            Px,
                        ),
                        "display": Keyword(
                            "block",
                        ),
                        "height": Length(
                            100.0,
                            Px,
                        ),
                        "border-width": Length(
                            4.0,
                            Px,
                        ),
                        "border-color": ColorValue(
                            Color {
                                r: 204,
                                g: 0,
                                b: 0,
                                a: 255,
                            },
                        ),
                    },
                    children: [
                        StyledNode {
                            node: Node {
                                children: [],
                                node_type: Text(
                                    "Hello, ",
                                ),
                            },
                            specified_values: {},
                            children: [],
                        },
                        StyledNode {
                            node: Node {
                                children: [
                                    Node {
                                        children: [],
                                        node_type: Text(
                                            "world!",
                                        ),
                                    },
                                ],
                                node_type: Element(
                                    ElementPreproty {
                                        tag_name: "span",
                                        attributes: {
                                            "id": "name",
                                        },
                                    },
                                ),
                            },
                            specified_values: {
                                "background": Keyword(
                                    "red",
                                ),
                                "display": Keyword(
                                    "inline",
                                ),
                                "color": Keyword(
                                    "white",
                                ),
                            },
                            children: [
                                StyledNode {
                                    node: Node {
                                        children: [],
                                        node_type: Text(
                                            "world!",
                                        ),
                                    },
                                    specified_values: {},
                                    children: [],
                                },
                            ],
                        },
                    ],
                },
                StyledNode {
                    node: Node {
                        children: [
                            Node {
                                children: [],
                                node_type: Text(
                                    "Goodbye!\n    ",
                                ),
                            },
                        ],
                        node_type: Element(
                            ElementPreproty {
                                tag_name: "p",
                                attributes: {
                                    "class": "inner",
                                    "id": "bye",
                                },
                            },
                        ),
                    },
                    specified_values: {
                        "margin-bottom": Length(
                            20.0,
                            Px,
                        ),
                        "border-color": ColorValue(
                            Color {
                                r: 204,
                                g: 0,
                                b: 0,
                                a: 255,
                            },
                        ),
                        "border-width": Length(
                            4.0,
                            Px,
                        ),
                        "height": Length(
                            100.0,
                            Px,
                        ),
                        "width": Length(
                            500.0,
                            Px,
                        ),
                        "display": Keyword(
                            "block",
                        ),
                        "background": ColorValue(
                            Color {
                                r: 255,
                                g: 255,
                                b: 0,
                                a: 255,
                            },
                        ),
                    },
                    children: [
                        StyledNode {
                            node: Node {
                                children: [],
                                node_type: Text(
                                    "Goodbye!\n    ",
                                ),
                            },
                            specified_values: {},
                            children: [],
                        },
                    ],
                },
            ],
        },
    ],
}
